manifests/zone.pp

```diff
@@ -24,21 +25,45 @@
     Optional[Enum['yes', 'no', 'explicit']] $dns_notify = undef,
 ) {
 
+  $_contact = pick($contact, "root.${zone}.")
+
   $zonefilename = "${zonefilepath}/${filename}"
 
+  if $::dns::enable_views {
+    if $target_views == undef or size($target_views) == 0 {
+      warning('You seem to mix BIND views with global zones, which will probably fail')
+      $_target_views = ['_GLOBAL_']
+    } else {
+      $_target_views = $target_views
+    }
+  } else {
+    $_target_views = ['_GLOBAL_']
+  }
+
   if $zonetype == 'slave' {
     $_dns_notify = pick($dns_notify, 'no')
   } else {
     $_dns_notify = $dns_notify
   }
 
-  concat::fragment { "dns_zones+10_${zone}.dns":
-    target  => $::dns::publicviewpath,
-    content => template('dns/named.zone.erb'),
-    order   => "10-${zone}",
+  $_target_views.each |$view| {
+    $target = $view ? {
+      '_GLOBAL_' => $::dns::publicviewpath,
+      default    => "${::dns::viewconfigpath}/${view}.conf",
+    }
+
+    concat::fragment { "dns_zones+10_${view}_${title}.dns":
+      target  => $target,
+      content => template('dns/named.zone.erb'),
+      order   => "${view}-11-${zone}-1",
+    }
+
+    unless ($view == '_GLOBAL_' or defined(Dns::View[$view])) {
+      fail("Please define a dns::view '${view}' before using it as a dns::zone target")
+    }
   }
 
-  if $manage_file {
+  if $manage_file and !defined(File[$zonefilename]) {
```

Comment:
I'm still not happy about the use of defined here. How about adding a parameter ```use_file``` (or something like that) and changing the logic in ```named.zone.erb``` to ```<% if @manage_file || @use_file -%>```? Then ```use_file``` can default to false and get the same behaviour but allow for the desired behaviour.